### PR TITLE
[2024] Day 6: Guard Gallivant

### PR DESCRIPTION
**Pull request by Nish7**
- [x] Part 1 ⭐
- [x] Part 2 ⭐

### CODE REVIEW

File: 2024/day6/day6.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+		}
+	}
+	return current
+}
+
+func GetStartingPosition(grid [][]string) ([2]int, [2]int, error) {
+	for i := 0; i < len(grid); i++ {
+		for j := 0; j < len(grid[0]); j++ {
+			switch grid[i][j] {
+			case ">":
```
Comment (Nish7) [inline 2024/day6/day6.go:107]: these are uneccessary remove them

File: 2024/day6/day6.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	positions, isLoop := MoveGuard(position, direction, grid)
+	fmt.Println(len(positions), isLoop)
+	loopPositions := PutLoopObstacle(position, direction, positions, grid)
+	fmt.Println(loopPositions)
+
+}
+
+func PutLoopObstacle(startPosition [2]int, direction [2]int, positions [][2]int, grid [][]string) int {
```
Comment (Nish7) [inline 2024/day6/day6.go:32]: this is brute force approach. lmk if any better approach

File: 2024/day6/day6.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	return lp
+}
+
+func MoveGuard(position [2]int, direction [2]int, grid [][]string) ([][2]int, bool) {
+	// apprpach: start with the intial positiong and direction keep switching the directions by 90deg once you find a obstacle and end the while loop when gets out of board
+	i, j := position[0], position[1]
+
+	var visited [][2][2]int
+	var visitedDir [][2]int
+	visited = append(visited, [2][2]int{{i, j}, {direction[0], direction[1]}}) // tracks visited cells with direction
```
Comment (Nish7) [inline 2024/day6/day6.go:58]: could be improved with a map

File: 2024/day6/day6.go (hunk 2 above)
Comment (Nish7) [reply 2024/day6/day6.go:32]: Part 2 takes more than 2min to compute. crazt

Comment (thisisadityapatel) [reply 2024/day6/day6.go:32]: > this is brute force approach. lmk if any better approach

a suggestion, have like a function, which at each instance on the front checks if in the right direction of the th current position there exists a path in the previous iterations, if so add an obstacle in front of it, check for cycle and increment the count if the cycle exits, if the cycle exits, after that remove the obstacle.

Comment (thisisadityapatel) [reply 2024/day6/day6.go:32]: this ensures that is_cycle function does not run on every positional point in the position, basically avoiding brut force :)

Comment (Nish7) [reply 2024/day6/day6.go:32]: Open up a fix PR. Thanks 👍

Comment (thisisadityapatel) [reply 2024/day6/day6.go:32]: Nai

**Pull request merged**